pydocstringformatter/run.py

```diff
@@ -98,26 +84,148 @@ def format_file(self, filename: Path) -> bool:
                     _utils.generate_diff(
                         tokenize.untokenize(tokens),
-                        tokenize.untokenize(changed_tokens),
+                        tokenize.untokenize(formatted_tokens),
                         filename_str,
                     )
                 )
 
         return is_changed
 
-    def format_files(self, filepaths: list[Path]) -> bool:
-        """Format a list of files."""
+    def format_file_tokens(
+        self, tokens: list[tokenize.TokenInfo], filename: Path
+    ) -> tuple[list[tokenize.TokenInfo], bool]:
+        """Format a list of tokens.
+
+        tokens: List of tokens to format.
+        filename: Name of the file the tokens are from.
+
+        Returns:
+            A tuple containing [1] the formatted tokens in a list
+            and a boolean indicating if the tokens were changed.
+
+        Raises:
+            RuntimeError:
+                If the formatters are not able to get to a stable
+                result after 3 iterations.
+
+                It reports what formatters are still modifying the tokens.
+        """
+        formatted_tokens: list[tokenize.TokenInfo] = []
         is_changed = False
 
-        for file in filepaths:
-            is_changed = self.format_file(file) or is_changed
+        for index, tokeninfo in enumerate(tokens):
+            new_tokeninfo = tokeninfo
 
-        return is_changed
+            if _utils.is_docstring(new_tokeninfo, tokens[index - 1]):
+                new_tokeninfo, changers, stable = self.apply_formatters_twice(
+                    new_tokeninfo
+                )
+                is_changed = is_changed or bool(changers)
+
+                # Try again to get a stable result, skipping the formatters that
+                # didn't change the tokens in former passes
+                new_tokeninfo, changers, stable = self.apply_formatters_twice(
+                    new_tokeninfo, changers
+                )
+
+                if not stable:
+                    if len(changers) > 1:
+                        msg = "Conflicting formatters: " + ", ".join(changers)
+                    else:
+                        msg = "Formatter: " + changers[0]
+                        msg += "Is not able to make stable changes on:\n"
+                        msg += new_tokeninfo.string + "\n"
+
+                    if len(changers) == 2:
+                        diff = _compare_formatters(
+                            new_tokeninfo,
+                            self.enabled_formatters[changers[0]],
+                            self.enabled_formatters[changers[1]],
+                            str(filename),
+                        )
+
+                        msg += f"\n{diff}"
+
+                    raise RuntimeError(msg)
+
+            formatted_tokens.append(new_tokeninfo)
+
+        return formatted_tokens, is_changed
+
+    @property
+    def enabled_formatters(self) -> dict[str, _formatting.Formatter]:
```

Comment:
Is there a reason why we don't just generate this in the `__init__` of the class? Then we can remove the `property` and private attribute. We have access to `config` at some point in the `__init__`.

Reply:
Well I did this this way thinking about https://github.com/DanielNoord/pydocstringformatter/pull/29 where eventually the setting up and the running would need to be decoupled.

The private attribute is just an explicit cache for the function.
LMK if you want me to move it to `__init__` ... maybe we could do a "middle of the road approach", where I rename this to `get_enabled_formatters` and set its result to `self.enabled_formatters` during `__init__`

Reply:
Yeah I think the last suggestion makes sense. It's a bit too long to put in the `__init__` but I think we should use a helper function/method to set it during the `__init__`.

For #29 we could do:
1. load config
2. load additional formatters
3. load enabled formatters

Reply:
OK, I will get that in the next commit